### PR TITLE
core: remove inline_block_at_pos

### PR DESCRIPTION
**Pull request by superlopuh**
First PR removing the APIs that let you pass an index in, this one is not used in the framework

### CODE REVIEW

**Review by math-fehr — COMMENTED** [review]
Hmm, In one way, I would say yes because we sholudn't reason in terms of position.
But at the same time, how do you inline a block in an empty block? We don't have any way of doing this now with this PR, so we should probably add these features before we remove this?
The missing methods are probably `inline_block_at_block_begin` or `end`.
I'm not sure what MLIR does here as well? Maybe we can look there?

**Comment by tobiasgrosser** [comment]
We could have sth like in-line at operation? Or split block and then in-line at the beginning. The best is to check what mlir does. We should also check if we need such feature for chocopy.

**Comment by math-fehr** [comment]
So just looking at it, mlir has `inline_block_before` and `inline_block_after`.
They do not have `inline_block_at_block_begin` (and end), since they can just use the `end()` iterator :/
So maybe we should add it despite MLIR not having them (as we don't have a pythonic solution)?

**Comment by tobiasgrosser** [comment]
What do you mean by not having a pythonic solution?

**Comment by math-fehr** [comment]
So if you have an iterator in C++, you often have a `iterator::end()` function that you can call even if the iterator is empty.
However, in Python, this does not exist, so we can't just say : "inline_block_at_ops(ops.end())".

But maybe there is something I'm overlooking

**Comment by superlopuh** [comment]
hmm, one way to do this would be to have a nullable `existing_op` in `inline_block_after`, which would append always

**Comment by tobiasgrosser** [comment]
> So if you have an iterator in C++, you often have a `iterator::end()` function that you can call even if the iterator is empty. However, in Python, this does not exist, so we can't just say : "inline_block_at_ops(ops.end())".
> 
> But maybe there is something I'm overlooking

First, instead of 'at', should we use 'before' and 'after'.

Second, Python does not offer an .end() as an iterator. However, can we not just pass the operation itself as an iterator? Or what is the type of our iterator? I would expect sth like Block::inline_block_after_op(op: Operation).

**Comment by tobiasgrosser** [comment]
Or are you concerned about empty blocks?

**Comment by math-fehr** [comment]
> Second, Python does not offer an .end() as an iterator. However, can we not just pass the operation itself as an iterator? Or what is the type of our iterator? I would expect sth like `Block::inline_block_after_op(op: Operation)`.

Yes sorry, we should use `after` and `before` when we can, which we already have.
The problem comes when we want to inline a block in an empty block. In that case, we don't have any operation that we can pass to our `rewriter`.

> hmm, one way to do this would be to have a nullable existing_op in inline_block_after, which would append always

We need to pass the block it needs to be inserted in then, a `None` `Operation` would not give any information on where to inline the block. Which is why I think we should add a `inline_block_at_end(Block)` and `inline_block_at_begin(Block)`.

File: xdsl/rewriter.py [hunk 1]
```diff
@@ -63,6 +63,32 @@ def replace_op(
...
+                if (
+                    isinstance(operand, BlockArgument)
+                    and operand.block is extended_block
+                ):
+                    raise Exception(
+                        "Cannot inline block which has operations using "
+                        "the block arguments."
+                    )
+
+        ops = list(inlined_block.ops)
```
Comment (tobiasgrosser) [inline xdsl/rewriter.py:86]: Why do we convert this to a list? This is asymptotically bad, no?

**Review by tobiasgrosser — CHANGES_REQUESTED** [review]

File: xdsl/rewriter.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/rewriter.py:86]: We need to, because we have to update all operation parents!
That's not something we can avoid sadly, MLIR has the same issue, and moving `n` operations is always `O(n)`.

Comment (tobiasgrosser) [reply xdsl/rewriter.py:86]: We should at least write sth like
```
for op in inline_lock.ops:
  op.detach()
```

Comment (superlopuh) [reply xdsl/rewriter.py:86]: detach detaches from block and next/previous, so this would not work

**Review by tobiasgrosser — APPROVED** [review]
OK, that's fine than.

**Review by webmiche — APPROVED** [review]

**Review by math-fehr — APPROVED** [review]
Nice!
Should we also add a similar `inline_block_at_begin`?
Otherwise, we need some ugly logic to do this!
(I'm fine to do this in another PR, but I would prefer here to be honest).

File: xdsl/pattern_rewriter.py [hunk 2]
```diff
@@ -225,6 +225,21 @@ def erase_block_argument(self, arg: BlockArgument, safe_erase: bool = True) -> N
         self.has_done_action = True
         arg.block.erase_arg(arg, safe_erase=safe_erase)
 
+    def inline_block_at_end(self, block: Block, target_block: Block):
+        """
+        Move the block operations to a given position in another block.
+        This block should not be a parent of the block to move to, and both blocks
+        should be child of the matched operation.
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/pattern_rewriter.py]: - [ ] remove matched op comment
- [ ] add `inline_block_at_start`
